core/imds/src/test/java/software/amazon/awssdk/imds/internal/DefaultEc2MetadataAsyncClientTest.java

```diff
@@ -0,0 +1,337 @@
+            verify(exactly(1), getRequestedFor(urlPathEqualTo(AMI_ID_RESOURCE))
+                .withHeader(TOKEN_HEADER, equalTo("some-token")));
+        } catch (Exception e) {
+            fail("Unexpected exception while executing test", e);
+        }
+    }
+
+    @Test
+    public void get_multipleAsyncRequest_shouldCompleteSuccessfully() {
+        int totalRequests = 128;
+        stubFor(put(urlPathEqualTo(TOKEN_RESOURCE_PATH)).willReturn(
+            aResponse().withFixedDelay(200).withBody("some-token")));
+        for (int i = 0; i < totalRequests; i++) {
+            ResponseDefinitionBuilder responseStub = aResponse().withFixedDelay(300)
+                                                                .withStatus(200)
+                                                                .withBody("response::" + i);
+            stubFor(get(urlPathEqualTo(AMI_ID_RESOURCE + "/" + i)).willReturn(responseStub));
+        }
+        try (Ec2MetadataAsyncClient client =
+                 Ec2MetadataAsyncClient.builder()
+                                       .endpoint(URI.create("http://localhost:" + mockMetadataEndpoint.port()))
+                                       .tokenTtl(Duration.ofSeconds(1024))
+                                       .build())
+        {
+            List<CompletableFuture<MetadataResponse>> requests = Stream.iterate(0, x -> x + 1)
+                                                                       .map(i -> client.get(AMI_ID_RESOURCE + "/" + i))
+                                                                       .limit(totalRequests)
+                                                                       .collect(Collectors.toList());
+            CompletableFuture<List<MetadataResponse>> responses =
+                CompletableFuture.allOf(requests.toArray(new CompletableFuture[0]))
+                                 .thenApply(unusedVoid -> requests.stream()
+                                                                  .map(CompletableFuture::join)
+                                                                  .collect(Collectors.toList()));
+
+            List<MetadataResponse> resolvedResponses = responses.join();
+            for (int i = 0; i < totalRequests; i++) {
+                MetadataResponse response = resolvedResponses.get(i);
+                assertThat(response.asString()).isEqualTo("response::" + i);
+            }
+            verify(exactly(totalRequests), putRequestedFor(urlPathEqualTo(TOKEN_RESOURCE_PATH))
+                .withHeader(EC2_METADATA_TOKEN_TTL_HEADER, equalTo("1024")));
+            verify(exactly(totalRequests), getRequestedFor(urlPathMatching(AMI_ID_RESOURCE + "/" + "\\d+"))
+                .withHeader(TOKEN_HEADER, equalTo("some-token")));
+        } catch (Exception e) {
+            fail("Unexpected exception while executing test", e);
+        }
+    }
+
+    @Test
+    public void get_largeResponse_shouldSucceed() {
```

Comment:
Is this test necessary? It doesn't seem to be related to any logic in IMDS. Maybe it's more applicable to http client?

Reply:
this test actually allowed me to catch a problem where the response bytes were "chuncked" and I was not receiving all of the data, and I had to update my code. I think it is good to keep to test for that case.